### ISSUE TITLE
Cannot disable modules without recovery 

### ISSUE THREAD

**Issue by simymo** [topjohnwu/Magisk#1286]
I have a huawei Mate 20 (HMA-L29) which doesn't have a compatible twrp recovery, as a result, I installed Magisk using the patched .img file downloaded from this post and everything works fine.

Today I installed a bad Magisk module and got a bootloop after reboot, I flashed back my stock recovery file and it revived with Magisk disabled. But now my problem ls every time I enabled Magisk by flashing the patched img, it will enter bootloop mode again, even the Magisk Manager is uninstalled completely, and I have no way to disable the installed modules without root access or magisk enabled. (Neither could create a ".disable_magisk" file in /cache via adb shell with no root access)

Rather than putting a .disable_magisk in cache folder, I believe it will be more helpful if it can be temporary disabled by placing a file in some specified path in sdcard.

**Comment by A2HI2H** [comment]
Try with flashing stock boot image

**Comment by VR-25** [comment]
Due to file based encryption, /data/media/0 is scrambled by the time Magisk starts. Placing a file there is unreliable, unless you don't use FBE (unlikely when unrooted). A potential solution would be forcing core only mode with a key hold/press.

**Comment by topjohnwu** [comment]
I did try to implement bootloop prevention but it had caused some issue. I will try to make it more stable in the next release

**Comment by jancm** [comment]
Untested but a possible temporary hack:

Place a script in /data/adb/post-fs-data.d
Have the script check for a flag file in /sdcard/
i.e. [ -e /sdcard/disable-magisk ]&&touch /cache/.disable_magisk

If you get a boot loop you can use adb to write the flag file to /sdcard

post-fs-data.d runs before any modules are loaded so should run before the boot loop kicks in. The next attempted boot after that should act on the .disable_magisk file.

Like I said untested, definitely not the cleanest method and a total hack - but I think it should work until something better comes along. Better than having to wipe data.

**Comment by RaymondSchnyder** [comment]
What if you could place the .disable_magisk file to /system?
Yea yea, i know. The system gets tampered.
But at this point, this could help:
1. Detect if /system/.disable_magisk exists
2. if yes, enable core only mod

After that, the user is able to remove the troublemakers and can restore the systempartition from his backup, so its systemless again.

Is that an ugly workaround or an idea?
because wiping my /data is definietly not something i would rather do. 

Also, if i understood the documentation correctly, on A/B Devices, the file should be placed at /data/cache/.disable_magisk.
Is that the only place?

**Comment by osm0sis** [comment]
I think checking /persist for .disable_magisk would be a good addition for newer devices, since /cache is dead on newer devices but most of these will have /persist and it's unencrypted. GApps already accept /persist for their .gapps-config files as well.

/cache should also remain for older devices, so that's what will make this trickier, since in https://github.com/topjohnwu/Magisk/blob/master/native/jni/core/bootstages.cpp there are quite a number of references to DISABLEFILE (set in magisk.h), so @topjohnwu or someone with more C++ chops than me will need to take a look at it. :+1:

**Comment by topjohnwu** [comment]
Fixed here e31e68760205427f1683a04353faea5724bdae32

**Comment by zy-lee** [comment]
> Fixed here [e31e687](https://github.com/topjohnwu/Magisk/commit/e31e68760205427f1683a04353faea5724bdae32)

I am not familiar with magisk tool, and I can't run `magisk --remove-modules` directly in adb shell. How to remove modules in ADB shell without root according to this commit?

**Comment by reyqn** [comment]
My understanding was that while in bootloop situation (magisk installed), you should have access to the magisk command after running adb shell.
So technically you have to be rooted but even if you can't boot you can remove all the modules.

**Comment by bittwiddlers** [comment]
So I have a bootloop with magisk enabled after installing another module. I don't have a working twrp for the device at the moment as there isn't one for Android 10 on the Pixel 3 yet so can't fix anything from twrp

I can adb in while it's hanging and type "magisk --remove-modules". After a pause it will reboot but the problem doesn't appear to be fixed. Is there anything else I can try? I can get the phone to boot by using an unpatched kernel but any with magisk in it will result in it hanging

**Comment by skittles9823** [comment]
@bittwiddlers you can use adb to delete the modules causing the issue with `adb shell` then `su -c rm -rf /data/adb/modules/<module id>`

Or you can create a core only mode boot image. There are instructions to do so [here](https://forum.xda-developers.com/pixel-3a-xl/themes/magisk-modules-disabler-booting-magisk-t3976625/post80418753).

**Comment by bittwiddlers** [comment]
On the pixel I don't have access to anything under the /data directory when I connect to the phone through "adb shell"

But thanks for the pointer to how to build the kernel. I found one by the same author for my pixel 3 that was already built and used that to get magisk set up again

Thank you

**Comment by skittles9823** [comment]
@bittwiddlers yea you need to su while in adb shell to see the directories.

Glad you found a core only boot.img though.

**Comment by Inchworm333** [comment]
A good tip is that even if you are stuck in a boot loop (Google G with load bar underneath) you can still use adb and you should be able to use the magisk command.

**Comment by osm0sis** [comment]
Only if it's far enough along in the boot to boot need an insecure boot.img, and/or only if you enabled USB Debugging before your bootloop...

**Comment by gudenau** [comment]
There should really be a way to create a core only boot image, since I am stuck in a boot loop from a module.

Things I have tried:
rm -vfr /data/adb/module/[module id]
touch /data/.disable_magisk
touch /data/.disable_magisk
magisk --remove-modules
adb sideload MultiFunctions4Magisk-v1.0.3.4.1.5A0(Unity4.2)(3MF).zip

The only thing that lets me boot is using the stock Pixel 4 XL boot.img, since there is no TWRP available I am kinda screwed.

**Comment by p1r473** [comment]
Hi, I got myself in a pickle
Im on a Pixel 4 XL. I accidently installed EdXposed 0.4.5.1_beta instead of 0.4.6 and it caused a bootloop
I was able to get back into the OS by reflashing the latest OTA update, but I am left with a problem:
Any time I reflash Magisk, it recreates /data/adb/modules/riru-core and riru_edxposed_sandhook , the old versions that cause bootloop
No matter how many times I go in and delete these folders with ADB, the ygety recreated any time I reflash Magisk

How can I prevent Magisk from reinstalling these modules every boot? Please help!

**Comment by p1r473** [comment]
> There should really be a way to create a core only boot image, since I am stuck in a boot loop from a module.
> 
> Things I have tried:
> rm -vfr /data/adb/module/[module id]
> touch /data/.disable_magisk
> touch /data/.disable_magisk
> magisk --remove-modules
> adb sideload MultiFunctions4Magisk-v1.0.3.4.1.5A0(Unity4.2)(3MF).zip
> 
> The only thing that lets me boot is using the stock Pixel 4 XL boot.img, since there is no TWRP available I am kinda screwed.

I am in the exact same situation.
Im on a Pixel 4 XL. I accidently installed EdXposed 0.4.5.1_beta instead of 0.4.6 and it caused a bootloop
I was able to get back into the OS by reflashing the latest OTA update, but I am left with a problem:
Any time I reflash Magisk, it recreates /data/adb/modules/riru-core and riru_edxposed_sandhook , the old versions that cause bootloop
No matter how many times I go in and delete these folders with ADB, they get recreated any time I reflash Magisk

How can I prevent Magisk from reinstalling these modules every boot? Please help!

**Comment by p1r473** [comment]
> magisk --remove-modules

eboot bootloade

> There should really be a way to create a core only boot image, since I am stuck in a boot loop from a module.
> 
> Things I have tried:
> rm -vfr /data/adb/module/[module id]
> touch /data/.disable_magisk
> touch /data/.disable_magisk
> magisk --remove-modules
> adb sideload MultiFunctions4Magisk-v1.0.3.4.1.5A0(Unity4.2)(3MF).zip
> 
> The only thing that lets me boot is using the stock Pixel 4 XL boot.img, since there is no TWRP available I am kinda screwed.

Flash this boot.img: https://forum.xda-developers.com/pixel-4-xl/themes/magisk-modules-disabler-booting-magisk-t3990557
Then try the steps again.

**Comment by gudenau** [comment]
> > magisk --remove-modules
> 
> eboot bootloade
> 
> > There should really be a way to create a core only boot image, since I am stuck in a boot loop from a module.
> > Things I have tried:
> > rm -vfr /data/adb/module/[module id]
> > touch /data/.disable_magisk
> > touch /data/.disable_magisk
> > magisk --remove-modules
> > adb sideload MultiFunctions4Magisk-v1.0.3.4.1.5A0(Unity4.2)(3MF).zip
> > The only thing that lets me boot is using the stock Pixel 4 XL boot.img, since there is no TWRP available I am kinda screwed.
> 
> Flash this boot.img: https://forum.xda-developers.com/pixel-4-xl/themes/magisk-modules-disabler-booting-magisk-t3990557
> Then try the steps again.

THANK YOU

Now if only John would fix it himself.

**Comment by DoganM95** [comment]
@topjohnwu could you please add an option in Manager App below "Select and Patch a File" to patch a file with core-only functionality?

**Comment by osm0sis** [comment]
> @topjohnwu could you please add an option in Manager App below "Select and Patch a File" to patch a file with core-only functionality?
> 
> That would enable any magisk user to create a core-only boot image for their device for each firmware version to recover from bootloops and would save you lots of "it doesn't work"s :)

#1842 and #1989 are both up for @topjohnwu to evaluate and add a feature along those lines.

**Comment by hltdev8642** [comment]
> # tldr:
> If you have a non booting magisk, a twrp which doesnt want to decrypt your /data and you cant write in /data/cache, read this tldr
> 
> Preperations:
> Get yourself the stock boot.img, modify it with magisk (only manager needed, no root functions) and save it on your pc.
> 
> 1. boot twrp with fastboot boot
> 2. use adb shell and remove the content of /data/adb
> 3. place .disable_magisk in /data and /cache
> 4. Boot or flash your patched boot.img
> 5. Your phone should boot and if you open magisk, it should show a working magisk installation
> 6. A dialog appears asking for additional steps, i do not have an answer for that at the moment
> 
> # Question
> What if you could place the .disable_magisk file to /system?
> Yea yea, i know. The system gets tampered.
> But at this point, this could help:
> 
> 1. Detect if /system/.disable_magisk exists
> 2. if yes, enable core only mod
> 
> After that, the user is able to remove the troublemakers and can restore the system partition from his backup, so its systemless again.
> 
> Is that an ugly workaround or an idea?
> because wiping my /data is definietly not something i would rather do.
> 
> Also, if i understood the documentation correctly, on A/B Devices, the file should be placed at /data/cache/.disable_magisk.
> Is that the only place?
> 
> # Ugly Workaround notes
> EDIT: I found, maybe, a horrible workaround.
> If your magisk isnt booting and your storage is all encrypted:
> 
> 1. place .disable_magisk in /data and /cache
> 2. go into /data/adb
> 3. Remove the encrypted folders with a bit more content in it.
> 4. IMPORTANT: WITH FASTBOOT BOOT, Boot a boot.img manipulated by magisk (To create one, just boot into your device with the stock boot.img and create one there)
> 5. Wait till its booted
> 6. .... continued?
> 
> I can boot in magisk that way and get the message "Requires additional setup"
> But, if i do the additional steps, it wont boot ever again.
> Additional: Magisk is booted up, i get the green checks, but safety check says the CTS profile mismatches.
> 
> My next steps now would be, flashing the patched magisk image after i booted it and hoping it boots up again.
> 
> # First try: (Direct Install, recommended by the magisk author)
> * Doesnt work, bootloop.
> * Also direct booting the previous patched boot img doesnt work, magic
> * Trying now to flash my patched boot.img. Maybe this helps?
> * Nope, doesnt help. Bootloop.
> 
> What buggs me:
> Everytime i install magisk / flash magisk, i get the boot message "verity mode is set to disabled".
> I remember it was "logging" in the past.
> I googled around, but could not find an approach for my device to reenable that.
> 
> # Second try:
> * Instead of booting the patched boot.img on the first boot, i flashed it.
> * I could boot into system, everything looks normal.
> * Magisk still wants to do "additional steps".
> * Rebooting the phone to verify it works this way. I guess it will not and i have to do all steps again
> * Nope, it works, even after reboot.
> 
> PS: I invested now 3 - 4 hours of debugging into this, its 6am, i hope someone in the same situation can use this what i wrote here

Thank you so much!!! I was able to disable the module that was causing problems so easily this way. You saved me a huge rebooting my phone ordeal 👍

**Comment by OdinVex** [comment]
I could also use the ability to core-only patch a boot.img with a tablet I have. It has no Recovery and is encrypted but I can flash a aarch64 boot.img
